org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/SonarLintChangeListener.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * SonarLint for Eclipse
+ * Copyright (C) 2015-2016 SonarSource SA
+ * sonarlint@sonarsource.com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarlint.eclipse.core.internal;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.IResourceChangeEvent;
+import org.eclipse.core.resources.IResourceChangeListener;
+import org.eclipse.core.resources.IResourceDelta;
+import org.eclipse.core.resources.IResourceDeltaVisitor;
+import org.eclipse.core.runtime.CoreException;
+import org.sonarlint.eclipse.core.internal.jobs.AnalyzeProjectJob;
+import org.sonarlint.eclipse.core.internal.jobs.AnalyzeProjectRequest;
+import org.sonarlint.eclipse.core.internal.resources.SonarLintProject;
+
+public class SonarLintChangeListener implements IResourceChangeListener {
+
+  @Override
+  public void resourceChanged(IResourceChangeEvent event) {
+    if (event.getType() == IResourceChangeEvent.POST_CHANGE) {
+      final Map<IProject, Collection<IFile>> changedFilesPerProject = new HashMap<>();
+      try {
+        event.getDelta().accept(new IResourceDeltaVisitor() {
+
+          @Override
+          public boolean visit(IResourceDelta delta) throws CoreException {
+            if (delta.getResource().getProject() == null) {
+              // Workspace root
+              return true;
+            }
+            final SonarLintProject sonarProject = SonarLintProject.getInstance(delta.getResource().getProject());
+            if (!sonarProject.isAutoEnabled()) {
+              return false;
+            }
+            if (delta.getResource().getType() == IResource.FILE) {
+              IFile file = (IFile) delta.getResource();
+              if (delta.getKind() == IResourceDelta.CHANGED
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)